### PR TITLE
An alternate implementation of the dialog component

### PR DESCRIPTION
**Pull request by scottpfeiffer**
This is an alternate implementation of the dialog component. It changes the approach and/or fixes a few issues I ran into when using in an existing Visualforce application.  Some of which are:

### Memory Leak in SPA
There is no global map of previously used dialog instances kept in memory.

### Global Namespace
All of the javascript code resides in an IIFE and the global name of the IIFE's return value can easily be changed ("Dialog"). All CSS class start with a unique prefix. Since the Visualforce engine uses some of these names already this prevents any collisions (in Aloha, less likely a problem in Lightning).

### Asynchronous Flow
The `alert()` method invokes the callback function for complete asynchronous support (just like `confirm` and `prompt`).

### prompt() Support
Added a `prompt()` method to replace the native prompt dialogs.

### Localization
Externalized the button labels for "OK" and "Cancel" for the caller to pass in (e.g. get from Custom Labels).

### A11Y
Changed a few aria attributes to better represent the dialog

### Improved UX Design
Made the dialog fully styled with pure HTML and CSS; no images for borders, etc. It looks much closer to a LDS component now.

### DOM Cleanup
This version dynamically creates the DOM elements needed and also removes them with the dialog is closed.

### Testability
Added mechanism to listen for events (e.g. "close") that may make it easier to automate testing. The id for the dialog is unique and can be used in testing selectors as well.

### Modern Javascript / Cleanup
Some general housekeeping/cleanup on old code (e.g. IE11) no longer needed; use of new ES6 syntax such as classes, arrow functions, etc.

### CODE REVIEW

**Review by tlagithub — CHANGES_REQUESTED** [review]

File: force-app/main/default/pages/editAccountName.page [hunk 1]
```diff
@@ -17,7 +17,7 @@
     <apex:form id="theForm">
         <apex:inputField id="accountNameField" value="{!Account.Name}" />
         <br/>
-        <apex:commandButton id="save" action="{!save}" value="Save" onclick="event.preventDefault(); Dialog.showConfirm('Are you sure?', () => { validateAndSubmit(this) })" />
+        <apex:commandButton id="save" action="{!save}" value="Save" onclick="event.preventDefault(); Dialog.confirm('Are you sure?', {ok: 'OK', cancel: 'Cancel'}, () => { validateAndSubmit(this) })" />
```
[the remark this answers is not in the document]
Comment (scottpfeiffer) [reply force-app/main/default/pages/editAccountName.page]: Good catch! Looks like the previous example didn't account for the return value of the confirm dialog either, but I have added it now.

**Review by tlagithub — APPROVED** [review]
Thanks Scott for all the updates.
Let wait on the outcome of our meeting next week before merging.

File: README.md [hunk 2]
```diff
@@ -59,3 +59,7 @@ This project introduces `Dialog.alert()`, `Dialog.confirm()`, and `Dialog.prompt
 8. To see an example of a confirm dialog, rename the account and save your changes. A confirm modal dialog should appear.
 
 9. To see an example of an alert dialog, add a special characer like !@#$%^&*() to the account name and save your changes. An alert dialog should display to indicate an error.
+
+\
+\
+Thank you to [KittyGiraudel](https://github.com/KittyGiraudel/a11y-dialog) for your previous work on a pure javascript dialog!
```
Comment (tlagithub) [inline README.md:65]: This looks fine

**Review by tarora10 — APPROVED** [review]

**Review by bochulindra — APPROVED** [review]
This looks like a fantastic improvement over our original dialog. I can make general sense of what the new dialog is doing and it looks good. However there are so many features and polish that are above my expertise that I won't be able to see any gaps that may be there. With that said, the dialog provides the core functionality we want it too and that's good enough for me to go ahead and ship this.

File: force-app/main/default/pages/editAccountName.page [hunk 3]
```diff
@@ -28,12 +28,19 @@
         function validateAndSubmit(theButton) {
             var field = document.getElementById('{!$Component.theForm.accountNameField}');
             return new Promise((resolve, reject) => {                
-                if (isValidAccountName(field.value)) {                    
+                if (isValidAccountName(field.value)) {
+                    // In cases where the form associated with this button
+                    // has inputFields and we expecte to have
```
Comment (bochulindra) [inline force-app/main/default/pages/editAccountName.page:33]: nitpick: can you fix "expecte" to "expect"?

File: force-app/main/default/pages/editAccountName.page [hunk 4]
```diff
@@ -16,7 +16,7 @@
     <h1>Edit Account Name</h1><br/><br/>
     <apex:form id="theForm">
         <apex:inputField id="accountNameField" value="{!Account.Name}" />
-        <br/>
+        <br/> 
         <apex:commandButton id="save" action="{!save}" value="Save" onclick="event.preventDefault(); Dialog.confirm('Are you sure?', {ok: 'OK', cancel: 'Cancel'}, (ret) => { if(ret) validateAndSubmit(this); })" />
```
Comment (bochulindra) [inline force-app/main/default/pages/editAccountName.page:20]: After meeting with you, I'm confused with how this works when we're using an apex:commandButton. Does the action actually get called? For example what if you change the action to {!delete}. Does the record get deleted?

**Review by bochulindra — CHANGES_REQUESTED** [review]
I added a question about using apex:commandButton

File: force-app/main/default/pages/editAccountName.page (hunk 4 above)
Comment (tlagithub) [reply force-app/main/default/pages/editAccountName.page:25]: I updated the comments to cover the scenario in this question.

**Review by bochulindra — APPROVED** [review]

**Pull request merged**